src/bc.hpp

```diff
@@ -0,0 +1,123 @@
+//  \brief Custom boundary conditions for AthenaPK
+//
+// Computes reflecting boundary conditions using AthenaPK's cons variable pack.
+//========================================================================================
+
+#include "bvals/bvals.hpp"
+#include "mesh/meshblock.hpp"
+
+using parthenon::Real;
+
+void ReflectingInnerX3(std::shared_ptr<parthenon::MeshBlockData<Real>> &mbd, bool coarse);
+void ReflectingOuterX3(std::shared_ptr<parthenon::MeshBlockData<Real>> &mbd, bool coarse);
+
+// Function for checking boundary flags: is this a domain or internal bound?
+//
+inline bool IsDomainBound(parthenon::MeshBlock *pmb, parthenon::BoundaryFace face) {
+  return !(pmb->boundary_flag[face] == parthenon::BoundaryFlag::block ||
+           pmb->boundary_flag[face] == parthenon::BoundaryFlag::periodic);
+}
+
+// Get zones which are inside the physical domain, i.e. set by computation or MPI halo
+// sync, not by problem boundary conditions.
+//
+inline auto GetPhysicalZones(parthenon::MeshBlock *pmb, parthenon::IndexShape &bounds)
+    -> std::tuple<parthenon::IndexRange, parthenon::IndexRange, parthenon::IndexRange> {
+  return std::tuple<parthenon::IndexRange, parthenon::IndexRange, parthenon::IndexRange>{
+      parthenon::IndexRange{IsDomainBound(pmb, parthenon::BoundaryFace::inner_x1)
+                                ? bounds.is(parthenon::IndexDomain::interior)
+                                : bounds.is(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x1)
+                                ? bounds.ie(parthenon::IndexDomain::interior)
+                                : bounds.ie(parthenon::IndexDomain::entire)},
+      parthenon::IndexRange{IsDomainBound(pmb, parthenon::BoundaryFace::inner_x2)
+                                ? bounds.js(parthenon::IndexDomain::interior)
+                                : bounds.js(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x2)
+                                ? bounds.je(parthenon::IndexDomain::interior)
+                                : bounds.je(parthenon::IndexDomain::entire)},
+      parthenon::IndexRange{IsDomainBound(pmb, parthenon::BoundaryFace::inner_x3)
+                                ? bounds.ks(parthenon::IndexDomain::interior)
+                                : bounds.ks(parthenon::IndexDomain::entire),
+                            IsDomainBound(pmb, parthenon::BoundaryFace::outer_x3)
+                                ? bounds.ke(parthenon::IndexDomain::interior)
+                                : bounds.ke(parthenon::IndexDomain::entire)}};
+}
+
+enum class BCSide { Inner, Outer };
+enum class BCType { Outflow, Reflect };
+
+template <parthenon::CoordinateDirection DIR, BCSide SIDE, BCType TYPE>
+void ApplyBC(parthenon::MeshBlock *pmb, parthenon::VariablePack<Real> &q,
+             parthenon::IndexRange &nvar, const bool is_normal, const bool coarse) {
```

Comment:
Should we instead template on:
```suggestion
template <parthenon::BoundaryFace FACE, BCType TYPE>
void ApplyBC(parthenon::MeshBlock *pmb, parthenon::VariablePack<Real> &q,
             parthenon::IndexRange &nvar, const bool is_normal, const bool coarse) {
```
and remove the `BCSide` enum? Would this work with your other code?

Reply:
I'm not quite sure how this function is used/should be used. In this PR it looks like it's only used to apply reflecting boundary conditions. Don't all the other pgen's have outflow boundary conditions applied as the default? Unless I've misunderstood something.